R/listing_export.R

```diff
@@ -0,0 +1,164 @@
+#' Export as plain text with page break symbol
+#'
+#' @inheritParams paginate_listing
+#' @param lst Listing object
+#' @param file character(1). File to write.
+#' @param  page_type  character(1).   Name   of  a  page  type.   See
+#'     `page_types`.   Ignored when  `pg_width` and  `pg_height`
+#'     are set directly.
+#' @param  landscape logical(1). Should the  dimensions of `page_type`
+#'     be inverted  for landscape?  Defaults to  `FALSE`, ignored when
+#'     `pg_width` and `pg_height` are set directly.
+#' @param font_family character(1). Name of a font family. An error
+#'     will be thrown if the family named is not monospaced. Defaults
+#'     to Courier.
+#' @param font_size numeric(1). Font size, defaults to 12.
+#' @param pg_width numeric(1). Page width in inches.
+#' @param pg_height numeric(1). Page height in inches.
+#' @param hsep character(1).   Characters  to  repeat   to  create
+#'     header/body separator line.
+#' @param indent_size numeric(1). Indent size in characters. Ignored
+#' when `x` is already a MatrixPrintForm object in favor of information
+#' there.
+#' @param max_width integer(1), character(1) or NULL. Width that title
+#'     and   footer   (including   footnotes)  materials   should   be
+#'     word-wrapped to. If NULL, it is  set to the current print width
+#'     of the  session (`getOption("width")`). If set to `"auto"`,
+#'     the width of the table (plus any table inset) is used. Ignored
+#'     completely if `tf_wrap` is `FALSE`.
+#' @param tf_wrap logical(1). Should  the texts for  title, subtitle,
+#'     and footnotes be wrapped?
+#' @param paginate logical(1). Should \code{lst} be paginated before writing the file.
+#' Defaults to `TRUE` if any sort of page dimension is specified.
+#' @param \dots Passed directly to \code{\link{paginate_listing}}
+#' @param page_break character(1). Page break symbol (defaults to outputting \code{"\\s"}).
+#' @return \code{file} (this function is called for the side effect of writing the file.
+#'
+#' @note When specified, `font_size` is used *only* to determine pagination based
+#' on page dimensions. The written file is populated in raw ASCII text, which
+#' does not have the concept of font size.
+#'
+#' @export
+#'
+#'
+#' @examples
+#'
+#' dat <- ex_adae
+#' lsting <- as_listing(dat[1:25, ], key_cols = c("USUBJID", "AESOC")) %>%
+#'   add_listing_col("AETOXGR") %>%
+#'   add_listing_col("BMRKR1", format = "xx.x") %>%
+#'   add_listing_col("AESER / AREL", fun = function(df) paste(df$AESER, df$AREL, sep = " / "))
+#' main_title(lsting) <- "this is some title"
+#' main_footer(lsting) <- "this is some footer"
+#' cat(export_as_txt(lsting, file = NULL, paginate = TRUE))
+#'
+export_as_txt <- function(lst, file = NULL,
```

Comment:
@clarkliming This could be a generic, though honestly it should be a function in formatters which invokes generics  along the lines of (pseudocode)

```
export_as_txt <- function(obj, file, page_type, ...) {
     paged <- paginate(obj, page_type, ...)
     
     unlink(file)
     lapply(paged, function(x) cat(toString(x), file = file, sep = page_break, append = TRUE)
}
```

The thing missing for the above to work is that `paginate` should be a generic, rather than a differently named function for each package (ie `paginate_table` and `paginate_listing`) Easy to do, but it is an API change so we'd need to do a deprecation cycle of the old versions...

Reply:
let's refactor this in the next PI, and move things from rtables, rlistings to formatters